core/trino-main/src/main/java/io/trino/sql/planner/AdaptivePlanner.java

```diff
@@ -0,0 +1,333 @@
+
+        PlanNode initialPlan = getInitialPlan(currentAdaptivePlan);
+        PlanNode currentPlan = getCurrentPlan(currentAdaptivePlan);
+
+        PlanNodeIdAllocator idAllocator = new PlanNodeIdAllocator(getMaxPlanId(currentPlan) + 1);
+        TableStatsProvider tableStatsProvider = new CachingTableStatsProvider(plannerContext.getMetadata(), session);
+        PlanNode optimizedPlan = optimizePlan(currentPlan, symbolAllocator, tableStatsProvider, stageRuntimeInfoProvider, idAllocator);
+
+        if (isPlanEqual(currentPlan, optimizedPlan)) {
+            return root;
+        }
+
+        PlanNode adaptivePlan = addAdaptivePlanNode(idAllocator, initialPlan, optimizedPlan, symbolAllocator.getTypes());
+        // validate the adaptive plan
+        try (var ignored = scopedSpan(plannerContext.getTracer(), "validate-adaptive-plan")) {
+            planSanityChecker.validateAdaptivePlan(adaptivePlan, session, plannerContext, typeAnalyzer, symbolAllocator.getTypes(), warningCollector);
+        }
+        return createSubPlans(adaptivePlan, symbolAllocator, fragmentIdAllocator, subPlans);
+    }
+
+    private SubPlan createSubPlans(
+            PlanNode planNode,
+            SymbolAllocator symbolAllocator,
+            PlanFragmentIdAllocator fragmentIdAllocator,
+            List<SubPlan> sourceSubPlans)
+    {
+        return planFragmenter.createSubPlans(
+                session,
+                fragmentIdAllocator,
+                new Plan(planNode, symbolAllocator.getTypes(), StatsAndCosts.empty()),
+                new PartitioningScheme(Partitioning.create(SINGLE_DISTRIBUTION, ImmutableList.of()), planNode.getOutputSymbols()),
+                sourceSubPlans,
+                false,
+                warningCollector);
+    }
+
+    private PlanNode optimizePlan(
+            PlanNode plan,
+            SymbolAllocator symbolAllocator,
+            TableStatsProvider tableStatsProvider,
+            StageRuntimeInfoProvider stageRuntimeInfoProvider,
+            PlanNodeIdAllocator idAllocator)
+    {
+        for (PlanOptimizer optimizer : planOptimizers) {
+            plan = optimizer.optimize(
+                    plan,
+                    new PlanOptimizer.Context(
+                            session,
+                            symbolAllocator.getTypes(),
+                            symbolAllocator,
+                            idAllocator,
+                            warningCollector,
+                            planOptimizersStatsCollector,
+                            tableStatsProvider,
+                            stageRuntimeInfoProvider));
+        }
+        return plan;
+    }
+
+    private PlanNode addAdaptivePlanNode(
+            PlanNodeIdAllocator idAllocator,
+            PlanNode initialPlan,
+            PlanNode optimizedPlanNode,
+            TypeProvider types)
+    {
+        if (initialPlan.getId() != optimizedPlanNode.getId()) {
```

Comment:
Currently no planner is required to generate a new id for "updated" plan node, but I'm not sure what else we could use. @losipiuk

Reply:
nit: One option might be to just keep full version of updated plan (wo) need for `AdaptivePlanNode`, but I'm not sure if it would simplify stuff very much